File: apps/zerg/backend/zerg/services/shipper/spool.py
"""Pointer-based offline spool for shipper resilience.

When the Zerg API is unreachable, byte-range pointers are queued locally
and replayed when connectivity is restored. NO payload storage -- source
files are re-read on retry.
"""

from __future__ import annotations

import logging
import sqlite3
from dataclasses import dataclass
from datetime import datetime
from datetime import timedelta
from datetime import timezone
from pathlib import Path

logger = logging.getLogger(__name__)

# Shared DB path and connection management
_DEFAULT_DB_NAME = "longhouse-shipper.db"
MAX_QUEUE_SIZE = 10_000
BASE_BACKOFF_SECONDS = 5.0
MAX_BACKOFF_SECONDS = 3600  # 1 hour
MAX_RETRY_COUNT = 50
DEAD_AGE_DAYS = 7


def _get_db_path(db_path: Path | None = None, claude_config_dir: Path | None = None) -> Path:
    """Resolve DB path from explicit path, config dir, or environment."""
    if db_path is not None:
        return db_path
    if claude_config_dir is None:
        import os

        config_dir = os.getenv("CLAUDE_CONFIG_DIR")
        claude_config_dir = Path(config_dir) if config_dir else Path.home() / ".claude"
    return claude_config_dir / _DEFAULT_DB_NAME


def get_shared_connection(db_path: Path) -> sqlite3.Connection:
    """Create a WAL-mode connection for the shared shipper DB."""
    db_path.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(str(db_path), timeout=10)
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA synchronous=NORMAL")
    conn.execute("PRAGMA busy_timeout=5000")
    return conn


def init_schema(conn: sqlite3.Connection) -> None:
    """Create tables if they don't exist."""
    conn.executescript("""
        CREATE TABLE IF NOT EXISTS file_state (
            path TEXT PRIMARY KEY,
            provider TEXT NOT NULL,
            queued_offset INTEGER NOT NULL DEFAULT 0,
            acked_offset INTEGER NOT NULL DEFAULT 0,
            session_id TEXT,
            provider_session_id TEXT,
            last_updated TEXT NOT NULL
        );

        CREATE TABLE IF NOT EXISTS spool_queue (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            provider TEXT NOT NULL,
            file_path TEXT NOT NULL,
            start_offset INTEGER NOT NULL,
            end_offset INTEGER NOT NULL,
            session_id TEXT,
            created_at TEXT NOT NULL,
            retry_count INTEGER DEFAULT 0,
            next_retry_at TEXT NOT NULL,
            last_error TEXT,
            status TEXT DEFAULT 'pending'
        );

        CREATE INDEX IF NOT EXISTS idx_spool_status
        ON spool_queue(status, next_retry_at);
    """)


@dataclass
class SpoolEntry:
    """A pointer entry in the spool queue."""

    id: int
    provider: str
    file_path: str
    start_offset: int
    end_offset: int
    session_id: str | None
    created_at: datetime
    retry_count: int
    last_error: str | None


class OfflineSpool:
    """Pointer-based queue for offline resilience.

    Stores file path + byte range pointers instead of payloads.
    Source files are re-read and re-parsed on retry.
    """

    def __init__(
        self,
        db_path: Path | None = None,
        claude_config_dir: Path | None = None,
        conn: sqlite3.Connection | None = None,
    ):
        self.db_path = _get_db_path(db_path, claude_config_dir)
        if conn is not None:
            self._conn = conn
        else:
            self._conn = get_shared_connection(self.db_path)
            init_schema(self._conn)

    @property
    def conn(self) -> sqlite3.Connection:
        return self._conn

    def close(self) -> None:
        self._conn.close()

    def enqueue(
        self,
        provider: str,
        file_path: str,
        start_offset: int,
        end_offset: int,
        session_id: str | None = None,
    ) -> bool:
        """Add a byte-range pointer to the spool.

        Returns False if queue is at capacity (backpressure).
        """
        if self.total_size() >= MAX_QUEUE_SIZE:
            logger.warning(f"Spool at capacity ({MAX_QUEUE_SIZE}), dropping enqueue for {file_path}")
            return False

        now = datetime.now(timezone.utc).isoformat()
        self._conn.execute(
            """INSERT INTO spool_queue (provider, file_path, start_offset, end_offset, session_id, created_at, next_retry_at, status)
               VALUES (?, ?, ?, ?, ?, ?, ?, 'pending')""",
            (provider, file_path, start_offset, end_offset, session_id, now, now),
        )
        self._conn.commit()
        return True

    def dequeue_batch(self, limit: int = 100) -> list[SpoolEntry]:
        """Get pending entries ready for retry."""
        now = datetime.now(timezone.utc).isoformat()
        cursor = self._conn.execute(
            """SELECT id, provider, file_path, start_offset, end_offset, session_id, created_at, retry_count, last_error
               FROM spool_queue
               WHERE status = 'pending' AND next_retry_at <= ?
               ORDER BY created_at ASC
               LIMIT ?""",
            (now, limit),
        )
        results = []
        for row in cursor.fetchall():
            created = row[6]
            if created.endswith("Z"):
                created = created[:-1] + "+00:00"
            results.append(
                SpoolEntry(
                    id=row[0],
                    provider=row[1],
                    file_path=row[2],
                    start_offset=row[3],
                    end_offset=row[4],
                    session_id=row[5],
                    created_at=datetime.fromisoformat(created),
                    retry_count=row[7],
                    last_error=row[8],
                )
            )
        return results

    def mark_shipped(self, entry_id: int) -> None:
        """Remove a successfully shipped entry."""
        self._conn.execute("DELETE FROM spool_queue WHERE id = ?", (entry_id,))
        self._conn.commit()

    def mark_failed(self, entry_id: int, error: str, max_retries: int = MAX_RETRY_COUNT) -> bool:
        """Increment retry count with exponential backoff.

        Returns True if entry is now permanently dead.
        """
        # Get current retry count
        cursor = self._conn.execute("SELECT retry_count FROM spool_queue WHERE id = ?", (entry_id,))
        row = cursor.fetchone()
        if not row:
            return True

        new_count = row[0] + 1

        if new_count >= max_retries:
            self._conn.execute(
                "UPDATE spool_queue SET status = 'dead', retry_count = ?, last_error = ? WHERE id = ?",
                (new_count, error, entry_id),
            )
            self._conn.commit()
            logger.warning(f"Spool entry {entry_id} dead after {new_count} retries: {error}")
            return True

        # Exponential backoff: min(base * 2^retry_count, max_backoff)
        backoff = min(BASE_BACKOFF_SECONDS * (2**new_count), MAX_BACKOFF_SECONDS)
        next_retry = (datetime.now(timezone.utc) + timedelta(seconds=backoff)).isoformat()

        self._conn.execute(
            "UPDATE spool_queue SET retry_count = ?, last_error = ?, next_retry_at = ? WHERE id = ?",
            (new_count, error, next_retry, entry_id),
        )
        self._conn.commit()
        return False

    def pending_count(self) -> int:
        """Count of pending entries."""
        cursor = self._conn.execute("SELECT COUNT(*) FROM spool_queue WHERE status = 'pending'")
        return cursor.fetchone()[0]

    def total_size(self) -> int:
        """Total entries (for backpressure check)."""
        cursor = self._conn.execute("SELECT COUNT(*) FROM spool_queue WHERE status IN ('pending', 'dead')")
        return cursor.fetchone()[0]

    def cleanup(self) -> int:
        """Remove dead entries older than 7 days and stale pending entries older than 7 days."""
        cutoff = (datetime.now(timezone.utc) - timedelta(days=DEAD_AGE_DAYS)).isoformat()
        cursor = self._conn.execute(
            "DELETE FROM spool_queue WHERE (status = 'dead' AND created_at < ?) OR (status = 'pending' AND created_at < ?)",
            (cutoff, cutoff),
        )
        count = cursor.rowcount
        self._conn.commit()
        if count > 0:
            logger.info(f"Cleaned up {count} old spool entries")
        return count

    def clear(self) -> None:
        """Clear all spool entries."""
        self._conn.execute("DELETE FROM spool_queue")
        self._conn.commit()

File: apps/engine/src/state/spool.rs
//! Pointer-based offline spool for retry resilience.
//!
//! Stores byte-range pointers (NOT payloads) into source files.
//! On retry, the source file is re-read and re-parsed.
//! Max queue size: 10,000 entries (backpressure).

use anyhow::Result;
use chrono::{DateTime, Utc};
use rusqlite::Connection;

/// Maximum spool entries before backpressure kicks in.
const MAX_QUEUE_SIZE: usize = 10_000;

/// Base backoff in seconds.
const BACKOFF_BASE: f64 = 5.0;

/// Maximum backoff in seconds (1 hour).
const BACKOFF_MAX: f64 = 3600.0;

/// Default max retries before marking dead.
const DEFAULT_MAX_RETRIES: u32 = 50;

/// A spool entry — pointer to a byte range in a source file.
#[derive(Debug, Clone)]
pub struct SpoolEntry {
    pub id: i64,
    pub provider: String,
    pub file_path: String,
    pub start_offset: u64,
    pub end_offset: u64,
    pub session_id: Option<String>,
    pub created_at: DateTime<Utc>,
    pub retry_count: u32,
    pub last_error: Option<String>,
}

/// Spool operations on a shared SQLite connection.
pub struct Spool<'a> {
    conn: &'a Connection,
}

impl<'a> Spool<'a> {
    pub fn new(conn: &'a Connection) -> Self {
        Self { conn }
    }

    /// Enqueue a byte-range pointer. Returns false if at capacity.
    pub fn enqueue(
        &self,
        provider: &str,
        file_path: &str,
        start_offset: u64,
        end_offset: u64,
        session_id: Option<&str>,
    ) -> Result<bool> {
        if self.total_size()? >= MAX_QUEUE_SIZE {
            tracing::warn!("Spool at capacity ({} entries), rejecting enqueue", MAX_QUEUE_SIZE);
            return Ok(false);
        }

        let now = Utc::now().to_rfc3339();
        self.conn.execute(
            "INSERT INTO spool_queue (provider, file_path, start_offset, end_offset, session_id, created_at, next_retry_at, status)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?6, 'pending')",
            rusqlite::params![
                provider,
                file_path,
                start_offset as i64,
                end_offset as i64,
                session_id,
                now,
            ],
        )?;
        Ok(true)
    }

    /// Get pending entries ready for retry (next_retry_at <= now).
    pub fn dequeue_batch(&self, limit: usize) -> Result<Vec<SpoolEntry>> {
        let now = Utc::now().to_rfc3339();
        let mut stmt = self.conn.prepare(
            "SELECT id, provider, file_path, start_offset, end_offset, session_id, created_at, retry_count, last_error
             FROM spool_queue
             WHERE status = 'pending' AND next_retry_at <= ?1
             ORDER BY created_at ASC
             LIMIT ?2",
        )?;
        let rows = stmt.query_map(rusqlite::params![now, limit as i64], |row| {
            Ok(SpoolEntry {
                id: row.get(0)?,
                provider: row.get(1)?,
                file_path: row.get(2)?,
                start_offset: row.get::<_, i64>(3)? as u64,
                end_offset: row.get::<_, i64>(4)? as u64,
                session_id: row.get(5)?,
                created_at: row
                    .get::<_, String>(6)
                    .map(|s| {
                        DateTime::parse_from_rfc3339(&s)
                            .map(|d| d.with_timezone(&Utc))
                            .unwrap_or_else(|_| Utc::now())
                    })?,
                retry_count: row.get::<_, i32>(7)? as u32,
                last_error: row.get(8)?,
            })
        })?;
        let mut result = Vec::new();
        for row in rows {
            result.push(row?);
        }
        Ok(result)
    }

    /// Remove a successfully shipped entry.
    pub fn mark_shipped(&self, entry_id: i64) -> Result<()> {
        self.conn
            .execute("DELETE FROM spool_queue WHERE id = ?", [entry_id])?;
        Ok(())
    }

    /// Mark entry as failed with exponential backoff. Returns true if now permanently dead.
    pub fn mark_failed(&self, entry_id: i64, error: &str) -> Result<bool> {
        self.mark_failed_with_max(entry_id, error, DEFAULT_MAX_RETRIES)
    }

    /// Mark failed with custom max retries.
    pub fn mark_failed_with_max(&self, entry_id: i64, error: &str, max_retries: u32) -> Result<bool> {
        // Get current retry count
        let retry_count: i32 = self.conn.query_row(
            "SELECT retry_count FROM spool_queue WHERE id = ?",
            [entry_id],
            |row| row.get(0),
        )?;
        let new_count = retry_count + 1;

        if new_count as u32 >= max_retries {
            // Mark as dead
            self.conn.execute(
                "UPDATE spool_queue SET status = 'dead', retry_count = ?1, last_error = ?2
                 WHERE id = ?3",
                rusqlite::params![new_count, error, entry_id],
            )?;
            return Ok(true);
        }

        // Exponential backoff: min(5 * 2^retry, 3600)
        let backoff_secs = (BACKOFF_BASE * 2.0_f64.powi(new_count)).min(BACKOFF_MAX);
        let next_retry = Utc::now() + chrono::Duration::seconds(backoff_secs as i64);

        self.conn.execute(
            "UPDATE spool_queue SET retry_count = ?1, last_error = ?2, next_retry_at = ?3
             WHERE id = ?4",
            rusqlite::params![new_count, error, next_retry.to_rfc3339(), entry_id],
        )?;
        Ok(false)
    }

    /// Count pending (retryable) entries.
    pub fn pending_count(&self) -> Result<usize> {
        let count: i64 = self.conn.query_row(
            "SELECT COUNT(*) FROM spool_queue WHERE status = 'pending'",
            [],
            |row| row.get(0),
        )?;
        Ok(count as usize)
    }

    /// Total entries (for backpressure check).
    pub fn total_size(&self) -> Result<usize> {
        let count: i64 = self.conn.query_row(
            "SELECT COUNT(*) FROM spool_queue",
            [],
            |row| row.get(0),
        )?;
        Ok(count as usize)
    }

    /// Remove dead entries older than 7 days. Returns count removed.
    pub fn cleanup(&self) -> Result<usize> {
        let cutoff = (Utc::now() - chrono::Duration::days(7)).to_rfc3339();
        let deleted = self.conn.execute(
            "DELETE FROM spool_queue WHERE status = 'dead' AND created_at < ?",
            [&cutoff],
        )?;
        // Also clean pending entries older than 7 days
        let deleted2 = self.conn.execute(
            "DELETE FROM spool_queue WHERE status = 'pending' AND created_at < ?",
            [&cutoff],
        )?;
        Ok(deleted + deleted2)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::db::open_db;

    fn setup() -> (tempfile::NamedTempFile, Connection) {
        let tmp = tempfile::NamedTempFile::new().unwrap();
        let conn = open_db(Some(tmp.path())).unwrap();
        (tmp, conn)
    }

    #[test]
    fn test_enqueue_dequeue() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        let ok = spool
            .enqueue("claude", "/path/a.jsonl", 0, 1000, Some("s1"))
            .unwrap();
        assert!(ok);
        assert_eq!(spool.pending_count().unwrap(), 1);

        let batch = spool.dequeue_batch(10).unwrap();
        assert_eq!(batch.len(), 1);
        assert_eq!(batch[0].file_path, "/path/a.jsonl");
        assert_eq!(batch[0].start_offset, 0);
        assert_eq!(batch[0].end_offset, 1000);
    }

    #[test]
    fn test_mark_shipped() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        spool.enqueue("claude", "/f", 0, 100, None).unwrap();
        let batch = spool.dequeue_batch(10).unwrap();
        spool.mark_shipped(batch[0].id).unwrap();
        assert_eq!(spool.pending_count().unwrap(), 0);
        assert_eq!(spool.total_size().unwrap(), 0);
    }

    #[test]
    fn test_mark_failed_backoff() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        spool.enqueue("claude", "/f", 0, 100, None).unwrap();
        let batch = spool.dequeue_batch(10).unwrap();
        let id = batch[0].id;

        // First failure — not dead yet
        let dead = spool.mark_failed(id, "connection refused").unwrap();
        assert!(!dead);
        assert_eq!(spool.pending_count().unwrap(), 1);

        // Entry should have retry_count = 1 and future next_retry_at
        let entry: (i32, String) = conn
            .query_row(
                "SELECT retry_count, next_retry_at FROM spool_queue WHERE id = ?",
                [id],
                |row| Ok((row.get(0)?, row.get(1)?)),
            )
            .unwrap();
        assert_eq!(entry.0, 1);
        // next_retry_at should be in the future
        let next: DateTime<Utc> = DateTime::parse_from_rfc3339(&entry.1)
            .unwrap()
            .with_timezone(&Utc);
        assert!(next > Utc::now());
    }

    #[test]
    fn test_mark_failed_dead_after_max() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        spool.enqueue("claude", "/f", 0, 100, None).unwrap();
        let batch = spool.dequeue_batch(10).unwrap();
        let id = batch[0].id;

        // Fail 3 times with max_retries=3
        for i in 0..3 {
            let dead = spool
                .mark_failed_with_max(id, &format!("err {}", i), 3)
                .unwrap();
            if i < 2 {
                assert!(!dead);
            } else {
                assert!(dead);
            }
        }

        // Should now be dead, not pending
        assert_eq!(spool.pending_count().unwrap(), 0);
        assert_eq!(spool.total_size().unwrap(), 1); // still in DB as dead
    }

    #[test]
    fn test_backpressure() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        // Fill to capacity (use a smaller number for testing by checking total_size)
        // We won't actually insert 10K rows — just verify the check works
        // Insert one and check that enqueue returns true
        let ok = spool.enqueue("claude", "/f", 0, 100, None).unwrap();
        assert!(ok);
    }

    #[test]
    fn test_cleanup() {
        let (_tmp, conn) = setup();
        let spool = Spool::new(&conn);

        // Insert a dead entry with old timestamp
        let old_date = (Utc::now() - chrono::Duration::days(10)).to_rfc3339();
        conn.execute(
            "INSERT INTO spool_queue (provider, file_path, start_offset, end_offset, created_at, next_retry_at, status)
             VALUES ('claude', '/old', 0, 100, ?1, ?1, 'dead')",
            [&old_date],
        )
        .unwrap();

        assert_eq!(spool.total_size().unwrap(), 1);
        let cleaned = spool.cleanup().unwrap();
        assert_eq!(cleaned, 1);
        assert_eq!(spool.total_size().unwrap(), 0);
    }
}
